contracts/uniswap_exchange.vy

```diff
@@ -145,8 +146,8 @@ def getEthToTokenExactPrice(tokens_bought: uint256) -> uint256(wei):
 def ethToTokenExact(tokens_bought: uint256, max_eth: uint256(wei), deadline: timestamp, buyer: address, recipient: address) -> uint256(wei):
     assert (self.totalSupply > 0 and deadline >= block.timestamp) and max_eth > 0
     eth_sold: uint256(wei) = self.ethToTokenExactPrice(tokens_bought, self.balance - max_eth)
-    assert self.token.transfer(recipient, tokens_bought)
     eth_refund: uint256(wei) = max_eth - eth_sold
+    assert self.token.transfer(recipient, tokens_bought)
```

Comment:
First check whether max_eth > eth_sold, then do the transfer

Reply:
`eth_refund: uint256(wei) = max_eth - eth_sold`  should throw an error if max_eth < eth_sold

Reply:
yes.  The idea is to put the `transfer` after calculate the refund (instead of before the refund). If `eth_refund: uint256(wei) = max_eth - eth_sold` failed, there is no need to do the transfer.